src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,636 @@
+        other => {
+            return Err(PyTypeError::new_err(format!(
+                "Invalid target basis: {other}"
+            )))
+        }
+    };
+    Ok(res)
+}
+
+const VALID_BASIS: [&str; 12] = [
+    "U321", "U3", "U", "PSX", "ZSX", "ZSXX", "U1X", "RR", "ZYZ", "ZXZ", "XYX", "XZX",
+];
+
+#[inline]
+fn angles_from_unitary(unitary: ArrayView2<Complex64>, target_basis: &str) -> [f64; 4] {
+    match target_basis {
+        "U321" => params_u3_inner(unitary),
+        "U3" => params_u3_inner(unitary),
+        "U" => params_u3_inner(unitary),
+        "PSX" => params_u1x_inner(unitary),
+        "ZSX" => params_u1x_inner(unitary),
+        "ZSXX" => params_u1x_inner(unitary),
+        "U1X" => params_u1x_inner(unitary),
+        "RR" => params_zyz_inner(unitary),
+        "ZYZ" => params_zyz_inner(unitary),
+        "ZXZ" => params_zxz_inner(unitary),
+        "XYX" => params_xyx_inner(unitary),
+        "XZX" => params_xzx_inner(unitary),
+        &_ => unreachable!(),
+    }
+}
+
+#[inline]
+fn compare_error_fn(
+    circuit: &OneQubitGateSequence,
+    error_map: &Option<&OneQubitGateErrorMap>,
+    qubit: usize,
+) -> (f64, usize) {
+    match error_map {
+        Some(global_err_map) => {
+            let err_map = &global_err_map.error_map[qubit];
+            let fidelity_product: f64 = circuit
+                .gates
+                .iter()
+                .map(|x| 1. - err_map.get(&x.0).unwrap_or(&0.))
+                .product();
+            (1. - fidelity_product, circuit.gates.len())
+        }
+        None => (circuit.gates.len() as f64, circuit.gates.len()),
+    }
+}
+
+fn compute_error(
+    gates: &[(String, Vec<f64>)],
+    error_map: Option<&OneQubitGateErrorMap>,
+    qubit: usize,
+) -> (f64, usize) {
+    match error_map {
+        Some(err_map) => {
+            let num_gates = gates.len();
+            let gate_fidelities: f64 = gates
+                .iter()
+                .map(|x| 1. - err_map.error_map[qubit].get(&x.0).unwrap_or(&0.))
+                .product();
+            (1. - gate_fidelities, num_gates)
+        }
+        None => (gates.len() as f64, gates.len()),
+    }
+}
+
+#[pyfunction]
+pub fn compute_error_one_qubit_sequence(
+    circuit: &OneQubitGateSequence,
+    qubit: usize,
+    error_map: Option<&OneQubitGateErrorMap>,
+) -> (f64, usize) {
+    compute_error(&circuit.gates, error_map, qubit)
+}
+
+#[pyfunction]
+pub fn compute_error_list(
+    circuit: Vec<(String, Vec<f64>)>,
+    qubit: usize,
+    error_map: Option<&OneQubitGateErrorMap>,
+) -> (f64, usize) {
+    compute_error(&circuit, error_map, qubit)
+}
+
+#[pyfunction]
+pub fn unitary_to_gate_sequence(
+    unitary: PyReadonlyArray2<Complex64>,
+    target_basis_list: Vec<&str>,
+    qubit: usize,
+    error_map: Option<&OneQubitGateErrorMap>,
+) -> PyResult<Option<OneQubitGateSequence>> {
+    for basis in &target_basis_list {
+        if !VALID_BASIS.contains(basis) {
```

Comment:
Since `VALID_BASIS` is only used once, isn't it better to move `VALID_BASIS` to a local scope? (I also use "bases" rather than "basis" here)
```rust
    const VALID_BASES: [&str; 12] = [
        "U321", "U3", "U", "PSX", "ZSX", "ZSXX", "U1X", "RR", "ZYZ", "ZXZ", "XYX", "XZX",
    ];
    for basis in &target_basis_list {
        if !VALID_BASES.contains(basis) {
```

Reply:
Sure I can move it into a local scope. When I put it in the module scope originally because I thought I'd need to use it in more than one place. But I ended up not needing that

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/9578/commits/a3be3c633008cca6473cf9d31b70d12cfa648ba9